dom/src/makeDOMDriver.ts

```diff
@@ -50,44 +51,99 @@ function reportSnabbdomError(err: any): void {
   (console.error || console.log)(err);
 }
 
+function makeDOMReady$(): Stream<null> {
+  return xs.create<null>({
+    start(lis: Listener<null>) {
+      if (document.readyState === 'loading') {
+        document.addEventListener('readystatechange', () => {
+          const state = document.readyState;
+          if (state === 'interactive' || state === 'complete') {
+            lis.next(null);
+            lis.complete();
+          }
+        });
+      } else {
+        lis.next(null);
+        lis.complete();
+      }
+    },
+    stop() {},
+  });
+}
+
 function makeDOMDriver(
   container: string | Element | DocumentFragment,
   options?: DOMDriverOptions,
 ): Driver<Stream<VNode>, MainDOMSource> {
   if (!options) {
     options = {};
   }
+  checkValidContainer(container);
   const modules = options.modules || defaultModules;
+  makeDOMDriverInputGuard(modules);
   const isolateModule = new IsolateModule();
   const patch = init([isolateModule.createModule()].concat(modules));
-  const rootElement = getValidNode(container) || document.body;
-  const vnodeWrapper = new VNodeWrapper(rootElement);
+  const domReady$ = makeDOMReady$();
+  let vnodeWrapper: VNodeWrapper;
   const delegators = new Map<string, EventDelegator>();
-  makeDOMDriverInputGuard(modules);
+  let mutationObserver: MutationObserver;
+  const mutationConfirmed$ = xs.create<null>({
+    start(listener) {
+      mutationObserver = new MutationObserver(() => listener.next(null));
+    },
+    stop() {
+      mutationObserver.disconnect();
+    },
+  });
 
   function DOMDriver(vnode$: Stream<VNode>, name = 'DOM'): MainDOMSource {
     domDriverInputGuard(vnode$);
     const sanitation$ = xs.create<null>();
-    const rootElement$ = xs
-      .merge(vnode$.endWhen(sanitation$), sanitation$)
-      .map(vnode => vnodeWrapper.call(vnode))
-      .fold(patch, toVNode(rootElement))
-      .drop(1)
-      .map(unwrapElementFromVNode)
-      .compose(dropCompletion) // don't complete this stream
-      .startWith(rootElement as any);
+
+    const firstRoot$ = domReady$.map(() => {
+      const firstRoot = getValidNode(container) || document.body;
+      vnodeWrapper = new VNodeWrapper(firstRoot);
+      return firstRoot;
+    });
+
+    // The mutation observer internal to mutationConfirmed$ should
+    // exist before elementAfterPatch$ calls mutationObserver.observe()
+    mutationConfirmed$.addListener({});
+
+    const elementAfterPatch$ = firstRoot$
```

Comment:
What is convoluted about this?
I believe scheduling is a solution to #592, but not for #699 (which this code right here is about), because #699 needs to be solved by *somehow* detecting if the DOM was mutated or not.

Reply:
Convoluted: `merge -> map -> fold -> drop -> map -> startWith -> map -> compose`, each with various degrees of complexity. It is hard to reason about. Smaller steps where the outputs of the various operations are assigned to well-named variables makes it much easier for anyone who didn't write this code to understand it. Either that or apply some comments that explain the _why_.

> I believe scheduling is a solution to #592, but not for #699 (which this code right here is about)

Fair enough.

Reply:
Sure, it's a long chain of operators. And yes, one can easily refactor that by just splitting it up and placing them in convenient helper functions. Has nothing to do with schedulers, though.

Good that I understood your comment is about code style, not about the actual fix.

Reply:
The scheduler mentioning _is_ related to this PR, as the PR references #592 explicitly. It may not have a _direct_ impact on this patch solution, but indirectly it definitely has. IMO, it is an underlying issue with xstream. For example, Most.js doesn’t emit during `Stream.run`, and I think that the same behavior applied to xstream would be the solution to all these issues. Just my two cents.

Of course, if this is just a temporary patch that will be addressed again later when #592 is handled, then I understand the approach.

Reply:
@Frikki, most.js also has scheduling problems: https://github.com/cujojs/most/issues/414
And adding a trampoling scheduler (like RxJS 4 had) has its downsides too. This PR has a comprehensive solution without introducing a scheduler.

Reply:
To that end, I do not agree with the solution. As you know, we are working intensively with Motorcycle and `@most/core`. We do _not_ have these issues. But it’s fine if Cycle.js and xstream want things implemented differently. After all, Cycle.js and Motorcycle are not the same, though they deal with the same concept. Thanks for the feedback.